qiskit/circuit/parametertable.py

```diff
@@ -12,43 +12,140 @@
 """
 Look-up table for variable parameters in QuantumCircuit.
 """
-import warnings
 import functools
-from collections.abc import MutableMapping, MappingView
+import warnings
+from collections.abc import MappingView, MutableMapping, Set
+from typing import MutableSequence
+
+
+class ParameterReferences(MutableSequence, Set):
+    """An (insertion) ordered set of instruction parameter slot references.
+    Items are expected in the form ``(instruction, param_index)``. Membership
+    testing is overriden such that items that are otherwise value-wise equal
+    are still considered distinct if their ``instruction``\\ s are referentially
+    distinct.
+
+    Items are ordered by insertion, and may be read or deleted by sequence index.
+    However, update by index and insert by index are not supported, since
+    these actions inherently violate insertion order.
+    """
+
+    def _instance_key(self, ref):
+        return (id(ref[0]), ref[1])
+
+    def __init__(self, *refs) -> None:
+        self._data = []
+        self._instance_ids = set()
+
+        for ref in refs:
+            if not isinstance(ref, tuple) or len(ref) != 2:
+                raise ValueError("refs must be in form (instruction, param_index)")
+            k = self._instance_key(ref)
+            if k not in self._instance_ids:
+                self._data.append(ref)
+                self._instance_ids.add(k)
+
+    @classmethod
+    def from_iterable(cls, iterable):
+        """Create a new instance from an iterable of ``(instruction, param_index)``
+        tuples."""
+        return cls(*iterable)
+
+    def __getitem__(self, index):
+        return self._data[index]
+
+    def __setitem__(self, index, ref):
+        raise NotImplementedError("Position is dictated by insertion order.")
+
+    def __delitem__(self, index):
+        ref = self._data[index]
+        del self._data[index]
+        self._instance_ids.discard(self._instance_key(ref))
+
+    def __len__(self):
+        return len(self._data)
+
+    def __iter__(self):
+        return iter(self._data)
+
+    def __contains__(self, x) -> bool:
+        return self._instance_key(x) in self._instance_ids
+
+    def __repr__(self) -> str:
+        return repr(self._data)
 
-from .instruction import Instruction
+    def insert(self, index, value) -> None:
+        raise NotImplementedError("Position is dictated by insertion order.")
```

Comment:
I'm not sure why you'd forbid this - it just causes you to violate the `MutableSequence` interface, and I'm not sure why it'd hurt to allow an override.  That said, I'm not sure if we need to maintain the `MutableSequence` part of this anyway.

Reply:
This is forbidden for the same reason as is set by index. I agree that it's less than ideal to violate `MutableSequence`. Perhaps we should get rid of index-based access from existing places in Terra and drop the seq-like property here.